File: src/main/java/com/aoranzhang/ezrentback/data/entity/Suite.java
package com.aoranzhang.ezrentback.data.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;
import org.hibernate.validator.constraints.NotEmpty;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Set;

@Entity
@Table(name = "suite")
public class Suite extends AbstractPersistentObject{

    @ManyToOne(fetch = FetchType.EAGER, cascade = CascadeType.ALL)
    @JoinColumn(name = "building", nullable = false)
    private Building building;

    @OneToMany(
            mappedBy = "suite",
            cascade = CascadeType.ALL,
            orphanRemoval = true,
            fetch = FetchType.EAGER
    )
    @JsonIgnore
    private Set<Room> rooms;

    @NotEmpty
    private String floor;

    @NotNull
    private Integer availability;

    @NotEmpty
    private String name;

    @NotNull
    private Integer rent;

    private boolean negotiable;

    public void copy(Suite suite) {
        if(suite.getId() != null && !suite.getId().equals("")) {
            setId(suite.getId());
        }
        setBuilding(suite.getBuilding());
        setAvailability(suite.getAvailability());
        setFloor(suite.getFloor());
        setName(suite.getName());
        setNegotiable(suite.isNegotiable());
        setRent(suite.getRent());
    }

    public Integer getRent() {
        return rent;
    }

    public void setRent(Integer rent) {
        this.rent = rent;
    }

    public boolean isNegotiable() {
        return negotiable;
    }

    public void setNegotiable(boolean negotiable) {
        this.negotiable = negotiable;
    }

    public Set<Room> getRooms() {
        return rooms;
    }

    public void setRooms(Set<Room> rooms) {
        this.rooms = rooms;
    }

    public Integer getAvailability() {
        return availability;
    }

    public void setAvailability(Integer availability) {
        this.availability = availability;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getFloor() {
        return floor;
    }

    public void setFloor(String floor) {
        this.floor = floor;
    }

    public Building getBuilding() {
        return building;
    }

    public void setBuilding(Building building) {
        this.building = building;
    }

}

File: src/main/kotlin/com/aoranzhang/ezrentback/data/entity/Suite.kt
package com.aoranzhang.ezrentback.data.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import com.google.common.collect.ImmutableSet

import javax.persistence.*
import javax.validation.constraints.NotNull

@Entity
@Table(name = "suite")
class Suite : AbstractPersistentObject {

    @ManyToOne(fetch = FetchType.EAGER, cascade = [CascadeType.ALL])
    @JoinColumn(name = "building", nullable = false)
    var building: Building? = null

    @OneToMany(mappedBy = "suite", cascade = [CascadeType.ALL], orphanRemoval = true, fetch = FetchType.EAGER)
    @JsonIgnore
    var rooms: Set<Room> = HashSet()

    var floor: String = ""

    @NotNull
    var available: Int = 0

    var name: String = ""

    @NotNull
    var rent: Int = 0

    var isRentNegotiable: Boolean = false

    constructor(): super() {}

    constructor(building: Building, rooms: Set<Room>, floor: String, available: Int = 1, name: String, rent: Int, isRentNegotiable: Boolean = true) : super() {
        this.building = building
        this.rooms = rooms
        this.floor = floor
        this.available = available
        this.name = name
        this.rent = rent
        this.isRentNegotiable = isRentNegotiable
    }

    constructor(suite: Suite): super(suite.id, suite.version) {
        this.building = suite.building
        this.rooms = suite.rooms
        this.floor = suite.floor
        this.available = suite.available
        this.name = suite.name
        this.rent = suite.rent
        this.isRentNegotiable = suite.isRentNegotiable
    }

    fun copy(suite: Suite) {
        this.building = suite.building
        this.rooms = suite.rooms
        this.floor = suite.floor
        this.available = suite.available
        this.name = suite.name
        this.rent = suite.rent
        this.isRentNegotiable = suite.isRentNegotiable
    }
}
